execution-engine/src/native_module_manager.rs

```diff
@@ -0,0 +1,291 @@
+        // this results in errors later, since the native module is not supposed
+        // to access these files
+        for f in &top_level_files {
+            let path = self.native_module_directory.join(strip_root_slash(&f));
+            let _ = create_dir(path);
+        }
+
+        Ok(top_level_files)
+    }
+
+    /// Recursively copy a `path` under the native module's directory to the
+    /// VFS.
+    /// Takes an unprefixed path, i.e. not including the path to the native
+    /// module's  directory.
+    /// Access errors are ignored.
+    /// This function should be called after the native module's execution to
+    /// reflect the side effects of execution onto the VFS.
+    fn copy_fs_to_vfs(&mut self, path_unprefixed: &Path) -> FileSystemResult<()> {
+        let path_prefixed = self.native_module_directory.join(strip_root_slash(&path_unprefixed));
+        if path_prefixed.is_dir() {
+            for entry in read_dir(path_prefixed)? {
+                let entry = entry?;
+                let path_prefixed = entry.path();
+                let path_unprefixed = path_prefixed.strip_prefix(&self.native_module_directory).map_err(|_| ErrNo::Access)?;
+
+                // Ignore execution configuration file
+                if path_unprefixed == PathBuf::from(EXECUTION_CONFIGURATION_FILE) {
+                    continue;
+                }
+
+                let path_unprefixed = PathBuf::from("/").join(path_unprefixed);
+                if path_prefixed.is_dir() {
+                    // Create directory on the VFS with `path_open()`
+                    let prestat = self.native_module_vfs.find_prestat(&path_unprefixed);
+                    if prestat.is_ok() {
+                        let (fd, file_name) = prestat?;
+                        self.native_module_vfs.path_open(
+                            fd,
+                            LookupFlags::empty(),
+                            file_name,
+                            OpenFlags::CREATE,
+                            FileSystem::DEFAULT_RIGHTS,
+                            FileSystem::DEFAULT_RIGHTS,
+                            FdFlags::empty(),
+                        )?;
+                        self.copy_fs_to_vfs(&path_unprefixed)?;
+                    }
+                } else {
+                    // Read file on the kernel fileystem, chunk by chunk
+                    let mut f = File::open(self.native_module_directory.join(&path_prefixed))?;
+                    let mut buf: [u8; 128] = [0; 128];
+
+                    // Copy it to the VFS. First truncate the VFS file first
+                    // then append to it. If the principal doesn't have write
+                    // access, just ignore it
+                    if self.native_module_vfs.write_file_by_absolute_path(&path_unprefixed, vec![], false).is_ok() {
+                        loop {
+                            let n = f.read(&mut buf)?;
+                            if n <= 0 {
```

Comment:
Minor: `n` is of type`usize` so cannot be negative.

Another way to write this without `break` command is `while f.read(&mut buf)? > 0 { ... }`

Reply:
We actually need to access `n` inside the loop